selectDynamic doesn't work with @targetName

## Compiler version

3.3.1

## Minimized code

```Scala
import scala.annotation.targetName

class X extends scala.reflect.Selectable:
	@targetName("hello")
	val `+` = "1"

val x = X()
x.selectDynamic("+")
```

## Output

```scala
java.lang.ExceptionInInitializerError
	at Main$.&lt;clinit&gt;(main.scala:14)
	at Main.main(main.scala)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at sbt.Run.invokeMain(Run.scala:144)
	at sbt.Run.execute$1(Run.scala:94)
	at sbt.Run.$anonfun$runWithLoader$5(Run.scala:121)
	at sbt.Run$.executeSuccess(Run.scala:187)
	at sbt.Run.runWithLoader(Run.scala:121)
	at sbt.Run.run(Run.scala:128)
	at com.olegych.scastie.sbtscastie.SbtScastiePlugin$$anon$1.$anonfun$run$1(SbtScastiePlugin.scala:38)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at sbt.util.InterfaceUtil$$anon$1.get(InterfaceUtil.scala:32)
	at sbt.ScastieTrapExit$App.run(ScastieTrapExit.scala:258)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.NoSuchMethodException: Playground$X.+()
	at java.base/java.lang.Class.getMethod(Class.java:2227)
	at scala.reflect.Selectable.applyDynamic(Selectable.scala:38)
	at scala.reflect.Selectable.applyDynamic$(Selectable.scala:11)
	at Playground$X.applyDynamic(main.scala:5)
	at scala.reflect.Selectable.selectDynamic(Selectable.scala:28)
	at scala.reflect.Selectable.selectDynamic$(Selectable.scala:11)
	at Playground$X.selectDynamic(main.scala:5)
	at Playground$.&lt;clinit&gt;(main.scala:10)
	... 17 more
```

## Expectation

It should resolve to the correct field independently of the presence of the `@targetName` annotation

--- a/compiler/src/dotty/tools/dotc/core/TypeComparer.scala
+++ b/compiler/src/dotty/tools/dotc/core/TypeComparer.scala
@@ -2084,7 +2084,7 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
             ExprType(info1.resType)
           case info1 => info1
 
-        isSubInfo(info1, info2, m.symbol.info.orElse(info1))
+        (isSubInfo(info1, info2, m.symbol.info.orElse(info1)) && m.symbol.hasTargetName(m.symbol.name))
         || matchAbstractTypeMember(m.info)
         || (tp1.isStable && m.symbol.isStableMember && isSubType(TermRef(tp1, m.symbol), tp2.refinedInfo))
       end qualifies

--- a/compiler/src/dotty/tools/dotc/core/TypeComparer.scala
+++ b/compiler/src/dotty/tools/dotc/core/TypeComparer.scala
@@ -2084,9 +2084,10 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
             ExprType(info1.resType)
           case info1 => info1
 
-        (isSubInfo(info1, info2, m.symbol.info.orElse(info1)) && m.symbol.hasTargetName(m.symbol.name))
-        || matchAbstractTypeMember(m.info)
-        || (tp1.isStable && m.symbol.isStableMember && isSubType(TermRef(tp1, m.symbol), tp2.refinedInfo))
+        m.symbol.hasTargetName(m.symbol.name) && (
+          isSubInfo(info1, info2, m.symbol.info.orElse(info1))
+          || matchAbstractTypeMember(m.info)
+          || (tp1.isStable && m.symbol.isStableMember && isSubType(TermRef(tp1, m.symbol), tp2.refinedInfo)))
       end qualifies
 
       tp1.member(name).hasAltWithInline(qualifies)

--- a/tests/neg/i18922.check
+++ b/tests/neg/i18922.check
@@ -1,0 +1,7 @@
+-- [E007] Type Mismatch Error: tests/neg/i18922.scala:11:27 ------------------------------------------------------------
+11 |def test = doClose(Resource()) // error
+   |                   ^^^^^^^^^^
+   |                   Found:    Resource
+   |                   Required: Object{def close(): Unit}
+   |
+   | longer explanation available when compiling with `-explain`

--- a/tests/neg/i18922.scala
+++ b/tests/neg/i18922.scala
@@ -1,0 +1,11 @@
+import scala.annotation.targetName
+
+def doClose(closable: { def close(): Unit }): Unit =
+  import reflect.Selectable.reflectiveSelectable
+  closable.close()
+
+class Resource:
+  @targetName("foo")
+  def close(): Unit = ???
+
+def test = doClose(Resource()) // error

--- a/tests/neg/targetName-refine.check
+++ b/tests/neg/targetName-refine.check
@@ -1,0 +1,7 @@
+-- [E007] Type Mismatch Error: tests/neg/targetName-refine.scala:7:27 --------------------------------------------------
+7 |val x: T { def f: Int } = C() // error
+  |                          ^^^
+  |                          Found:    C
+  |                          Required: T{def f: Int}
+  |
+  | longer explanation available when compiling with `-explain`

--- a/tests/neg/targetName-refine.scala
+++ b/tests/neg/targetName-refine.scala
@@ -4,5 +4,5 @@ trait T:
 class C extends T:
   @targetName("f2") def f: Int = 1
 
-val x: T { def f: Int } = C()
+val x: T { def f: Int } = C() // error
 

